### PR TITLE
Split Timeout into separate Timeout and Placeholder primitives

### PR DESCRIPTION
**Pull request by acdlite**
I think this separation makes sense, if for no other reason than it helps with the mental model. A Timeout determines the threshold before a suspend "effect" turns into an error. A Placeholder is like an error boundary that pattern matches on those errors.

Depends on https://github.com/facebook/react/pull/12279. Opening this separately, since it's not needed for the basic functionality and it's debatable whether these are the right primitives.

### CODE REVIEW

File: packages/shared/ReactSymbols.js [hunk 1]
```diff
@@ -39,6 +39,12 @@ export const REACT_ASYNC_MODE_TYPE = hasSymbol
 export const REACT_FORWARD_REF_TYPE = hasSymbol
   ? Symbol.for('react.forward_ref')
   : 0xead0;
+export const REACT_PLACEHOLDER_TYPE = hasSymbol
+  ? Symbol.for('react.placeholder')
+  : 0xead1;
+export const REACT_TIMEOUT_TYPE = hasSymbol
+  ? Symbol.for('react.timeout')
+  : 0xead2;
```
Comment (bvaughn) [inline packages/shared/ReactSymbols.js]: These new symbols should be added to `ReactIs` and `getComponentName`.

**Comment by react-linter** [comment]
React: size: :small_red_triangle:+1.0%, gzip: :small_red_triangle:+0.7%

ReactDOM: size: :small_red_triangle:+1.1%, gzip: :small_red_triangle:+1.2%


  <details>
  <summary>Details of bundled changes.</summary>

  <p>Comparing: ad7cd686670d8e519789fb226d8ff9175fb69370...271aa2f318012eb9eec7ae61ae263680bfedae8f</p>


  
## react
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react.development.js | +2.1% | +3.1% | 55.43 KB | 56.61 KB | 15.23 KB | 15.69 KB | UMD_DEV
**react.production.min.js** | **:small_red_triangle:+1.0%** | **:small_red_triangle:+0.7%** | **6.96 KB** | **7.03 KB** | **2.96 KB** | **2.98 KB** | **UMD_PROD**
react.development.js | +2.6% | +3.9% | 46.07 KB | 47.25 KB | 12.87 KB | 13.37 KB | NODE_DEV
**react.production.min.js** | **:small_red_triangle:+1.2%** | **:small_red_triangle:+0.8%** | **5.5 KB** | **5.57 KB** | **2.4 KB** | **2.42 KB** | **NODE_PROD**
React-dev.js | +2.7% | +3.6% | 46.07 KB | 47.33 KB | 12.55 KB | 13.01 KB | FB_WWW_DEV
**React-prod.js** | **:small_red_triangle:+0.1%** | **:small_red_triangle:+1.7%** | **13.38 KB** | **13.39 KB** | **3.71 KB** | **3.77 KB** | **FB_WWW_PROD**

## react-dom
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-dom.development.js | +3.0% | +2.6% | 611.8 KB | 630.2 KB | 141.4 KB | 145 KB | UMD_DEV
**react-dom.production.min.js** | **:small_red_triangle:+1.1%** | **:small_red_triangle:+1.2%** | **100.06 KB** | **101.14 KB** | **31.84 KB** | **32.23 KB** | **UMD_PROD**
react-dom.development.js | +3.1% | +2.7% | 596.18 KB | 614.57 KB | 137.23 KB | 140.92 KB | NODE_DEV
**react-dom.production.min.js** | **:small_red_triangle:+1.1%** | **:small_red_triangle:+1.0%** | **98.5 KB** | **99.55 KB** | **31.07 KB** | **31.37 KB** | **NODE_PROD**
react-dom-server.browser.development.js | 0.0% | 0.0% | 101.34 KB | 101.35 KB | 26.45 KB | 26.46 KB | UMD_DEV
react-dom-server.browser.development.js | 0.0% | 0.0% | 90.64 KB | 90.65 KB | 24.19 KB | 24.2 KB | NODE_DEV
react-dom-server.node.development.js | 0.0% | 0.0% | 92.56 KB | 92.57 KB | 24.74 KB | 24.74 KB | NODE_DEV
ReactDOM-dev.js | +3.1% | +2.7% | 620.62 KB | 639.72 KB | 139.99 KB | 143.78 KB | FB_WWW_DEV
**ReactDOM-prod.js** | **:small_red_triangle:+4.4%** | **:small_red_triangle:+3.6%** | **284.19 KB** | **296.57 KB** | **51.94 KB** | **53.8 KB** | **FB_WWW_PROD**
ReactDOMServer-dev.js | 0.0% | +0.1% | 94.11 KB | 94.16 KB | 24.03 KB | 24.04 KB | FB_WWW_DEV

## react-art
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-art.development.js | +4.4% | +4.0% | 414.71 KB | 433.1 KB | 90.31 KB | 93.93 KB | UMD_DEV
**react-art.production.min.js** | **:small_red_triangle:+1.2%** | **:small_red_triangle:+1.2%** | **90.38 KB** | **91.44 KB** | **27.49 KB** | **27.83 KB** | **UMD_PROD**
react-art.development.js | +5.4% | +5.1% | 340.55 KB | 358.94 KB | 71.63 KB | 75.29 KB | NODE_DEV
**react-art.production.min.js** | **:small_red_triangle:+1.9%** | **:small_red_triangle:+2.1%** | **54.9 KB** | **55.95 KB** | **16.77 KB** | **17.13 KB** | **NODE_PROD**
ReactART-dev.js | +5.5% | +5.3% | 348.81 KB | 367.91 KB | 71.21 KB | 74.99 KB | FB_WWW_DEV
**ReactART-prod.js** | **:small_red_triangle:+7.4%** | **:small_red_triangle:+6.8%** | **166.92 KB** | **179.21 KB** | **27.43 KB** | **29.31 KB** | **FB_WWW_PROD**

## react-test-renderer
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-test-renderer.development.js | +5.3% | +4.9% | 348.62 KB | 367 KB | 73.37 KB | 76.98 KB | UMD_DEV
**react-test-renderer.production.min.js** | **:small_red_triangle:+1.9%** | **:small_red_triangle:+1.8%** | **54.89 KB** | **55.95 KB** | **16.72 KB** | **17.01 KB** | **UMD_PROD**
react-test-renderer.development.js | +5.4% | +5.2% | 339.45 KB | 357.83 KB | 70.66 KB | 74.31 KB | NODE_DEV
**react-test-renderer.production.min.js** | **:small_red_triangle:+1.9%** | **:small_red_triangle:+2.1%** | **54.12 KB** | **55.17 KB** | **16.3 KB** | **16.64 KB** | **NODE_PROD**
ReactTestRenderer-dev.js | +5.5% | +5.4% | 347.98 KB | 367.07 KB | 70.31 KB | 74.07 KB | FB_WWW_DEV

## react-reconciler
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-reconciler.development.js | +5.7% | +5.5% | 319.8 KB | 338.19 KB | 66.12 KB | 69.77 KB | NODE_DEV
**react-reconciler.production.min.js** | **:small_red_triangle:+2.2%** | **:small_red_triangle:+2.6%** | **47 KB** | **48.05 KB** | **14.23 KB** | **14.6 KB** | **NODE_PROD**
react-reconciler-persistent.development.js | +5.8% | +5.5% | 319.14 KB | 337.51 KB | 65.88 KB | 69.53 KB | NODE_DEV
**react-reconciler-persistent.production.min.js** | **:small_red_triangle:+2.2%** | **:small_red_triangle:+2.0%** | **45.97 KB** | **47 KB** | **14.09 KB** | **14.37 KB** | **NODE_PROD**

## react-native-renderer
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
ReactNativeRenderer-dev.js | +4.1% | +3.9% | 459.6 KB | 478.6 KB | 98.34 KB | 102.15 KB | RN_FB_DEV
**ReactNativeRenderer-prod.js** | **:small_red_triangle:+5.8%** | **:small_red_triangle:+5.4%** | **217.88 KB** | **230.45 KB** | **36.4 KB** | **38.35 KB** | **RN_FB_PROD**
ReactNativeRenderer-dev.js | +4.1% | +3.9% | 459.35 KB | 478.33 KB | 98.28 KB | 102.09 KB | RN_OSS_DEV
**ReactNativeRenderer-prod.js** | **:small_red_triangle:+2.0%** | **:small_red_triangle:+1.8%** | **217.09 KB** | **221.52 KB** | **36.27 KB** | **36.91 KB** | **RN_OSS_PROD**
ReactFabric-dev.js | +4.3% | +4.1% | 442.01 KB | 461 KB | 93.94 KB | 97.75 KB | RN_FB_DEV
**ReactFabric-prod.js** | **:small_red_triangle:+2.2%** | **:small_red_triangle:+1.9%** | **202.77 KB** | **207.14 KB** | **33.67 KB** | **34.31 KB** | **RN_FB_PROD**
ReactFabric-dev.js | +4.3% | +4.1% | 442.05 KB | 461.03 KB | 93.95 KB | 97.76 KB | RN_OSS_DEV
**ReactFabric-prod.js** | **:small_red_triangle:+2.2%** | **:small_red_triangle:+1.9%** | **202.81 KB** | **207.18 KB** | **33.69 KB** | **34.32 KB** | **RN_OSS_PROD**

  </details>
  
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: packages/shared/ReactSymbols.js (hunk 1 above)
Comment (bvaughn) [reply packages/shared/ReactSymbols.js]: Probably also need a follow-up PR for DevTools to add the new type in so it doesn't show as "unknown"?

Comment (gaearon) [reply packages/shared/ReactSymbols.js]: Also test renderer 😄

**Comment by sebmarkbage** [comment]
We've been able to do a lot of refactors and change significantly how React works because nobody relies on subtle differences in practice. The APIs we exposed are really the highest level possible that we could get away with, not the lowest level. The lowest level used to be updateComponent or something on the object oriented API and would be beginWork or something in the new API. Exposing the lowest primitive is on its own not a virtue.

What's the use case that can't be described with just a Timeout and do you expect people to actually use these separately in practice?

**Comment by acdlite** [comment]
I think there are many places where you'd use a Placeholder without a Timeout. In fact, I think *usually* you would use Placeholder without overriding the timeout.

There are no places where you should use Timeout without a Placeholder. At least that I can think of.

So the reason to separate them is to avoid overuse of Timeout. I'm concerned if they're both in the same component, it'll be too tempting to specify a `delayMs` on every single placeholder. Which I guess isn't the worst thing in the world? Could lead to overcoupling but I suppose that's a factor either way.

Now that I type this out, a better solution might be to keep the two primitives combined but rename Timeout to Placeholder, like we do with the higher-level API we use in demos.

**Comment by sebmarkbage** [comment]
Yea, I think the only reason we named it Timeout was because we expected the higher-level thing to be called Placeholder and that you'd never use Timeout in practice.

Maybe that's just an indication that we do need the higher level thing to be built-in and we just have to solve the cross-renderer issue and what it means to hide something.

**Pull request closed without merging**